---
title: "Introduction to BioC Single-Cell Experiment"
author: "ks"
date: "3/22/2022"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# {.tabset}

## Data Infrastructure

Review this formatting of data, then run workflow  with PBMC data.
http://bioconductor.org/books/3.12/OSCA/data-infrastructure.html#background


```{r install-pkg}
if (!require("BiocManager", quietly = TRUE))
        install.packages("BiocManager")
if (!require("SingleCellExperiment"))        
        BiocManager::install('SingleCellExperiment')
if(!require('scater')){BiocManager::install('scater')} 
if(!require('scran')){BiocManager::install('scran')} 
if(!require('uwot')){BiocManager::install('uwot')} 
if(!require('batchelor')){BiocManager::install('batchelor')} 
if(!require('TENxPBMCData')){BiocManager::install('TENxPBMCData')} 
library(TENxPBMCData)
library(SingleCellExperiment)
```

## 3 Datasets

http://bioconductor.org/books/3.12/OSCA/filtered-human-pbmcs-10x-genomics.html

We're going to individually analyze 3 datasets that we combine in the end for a combined analysis.

```{r 10x}
library(TENxPBMCData)
all.sce <- list(
    pbmc3k=TENxPBMCData('pbmc3k'),
    pbmc4k=TENxPBMCData('pbmc4k'),
    pbmc8k=TENxPBMCData('pbmc8k')
)
```

##QC

Save unfiltered data.
```{r  all-data}
unfiltered <- all.sce
```

Filter on \% mt-DNA.

```{r filtered}
library(scater)
stats <- high.mito <- list()
for (n in names(all.sce)) {
    current <- all.sce[[n]]
    is.mito <- grep("MT", rowData(current)$Symbol_TENx)
    stats[[n]] <- perCellQCMetrics(current, subsets=list(Mito=is.mito))
    high.mito[[n]] <- isOutlier(stats[[n]]$subsets_Mito_percent, type="higher")
    all.sce[[n]] <- current[,!high.mito[[n]]]
}
```



```{r  qc-mtdna}
qcplots <- list()
for (n in names(all.sce)) {
    current <- unfiltered[[n]]
    colData(current) <- cbind(colData(current), stats[[n]])
    current$discard <- high.mito[[n]]
    qcplots[[n]] <- plotColData(current, x="sum", y="subsets_Mito_percent",
        colour_by="discard") + scale_x_log10()
}
do.call(gridExtra::grid.arrange, c(qcplots, ncol=3))
```

```{r low-quality}
lapply(high.mito, summary)
```

## Normalization

```{r lognorm}
all.sce <- lapply(all.sce, logNormCounts)
```

```{r sizefactors}
lapply(all.sce, function(x) summary(sizeFactors(x)))
```

##  Variance modeling

```{r hvg}
library(scran)
all.dec <- lapply(all.sce, modelGeneVar)
all.hvgs <- lapply(all.dec, getTopHVGs, prop=0.1)
```


```{r curve}
par(mfrow=c(1,3))
for (n in names(all.dec)) {
    curdec <- all.dec[[n]]
    plot(curdec$mean, curdec$total, pch=16, cex=0.5, main=n,
        xlab="Mean of log-expression", ylab="Variance of log-expression")
    curfit <- metadata(curdec)
    curve(curfit$trend(x), col='dodgerblue', add=TRUE, lwd=2)
}
```

## Dim Reduction

"For various reasons, we will first analyze each PBMC dataset separately rather than merging them together. We use randomized SVD, which is more efficient for file-backed matrices."

This takes a little while to run.

```{r  randomized-svd}
library(BiocSingular)
set.seed(10000)
all.sce <- mapply(FUN=runPCA, x=all.sce, subset_row=all.hvgs, 
    MoreArgs=list(ncomponents=25, BSPARAM=RandomParam()), 
    SIMPLIFY=FALSE)

set.seed(100000)
all.sce <- lapply(all.sce, runTSNE, dimred="PCA")

set.seed(1000000)
all.sce <- lapply(all.sce, runUMAP, dimred="PCA")
```

## Clustering

```{r cluster-cells}
for (n in names(all.sce)) {
    g <- buildSNNGraph(all.sce[[n]], k=10, use.dimred='PCA')
    clust <- igraph::cluster_walktrap(g)$membership
    colLabels(all.sce[[n]])  <- factor(clust)
}
```

```{r cluster-sizes}
lapply(all.sce, function(x) table(colLabels(x)))
```

##  Viz: tSNE

```{r viz-tsne}
all.tsne <- list()
for (n in names(all.sce)) {
    all.tsne[[n]] <- plotTSNE(all.sce[[n]], colour_by="label") + ggtitle(n)
}
do.call(gridExtra::grid.arrange, c(all.tsne, list(ncol=2)))
```

```{r viz-umap}
all.umap <- list()
for (n in names(all.sce)) {
    all.umap[[n]] <- plotUMAP(all.sce[[n]], colour_by="label") + ggtitle(n)
}
do.call(gridExtra::grid.arrange, c(all.umap, list(ncol=2)))
```

## Combine Experiments

Repeat the analysis after merging together the three batches.

```{r intersect-genes-batchcorrect}
# Intersecting the common genes.
universe <- Reduce(intersect, lapply(all.sce, rownames))
all.sce2 <- lapply(all.sce, "[", i=universe,)
all.dec2 <- lapply(all.dec, "[", i=universe,)

# Renormalizing to adjust for differences in depth.
library(batchelor)
normed.sce <- do.call(multiBatchNorm, all.sce2)

# Identifying a set of HVGs using stats from all batches.
combined.dec <- do.call(combineVar, all.dec2)
combined.hvg <- getTopHVGs(combined.dec, n=5000)

set.seed(1000101)
merged.pbmc <- do.call(fastMNN, c(normed.sce, 
    list(subset.row=combined.hvg, BSPARAM=RandomParam())))
```

The percentage of lost variance is reported as a diagnostic measure.

```{r lost.var}
metadata(merged.pbmc)$merge.info$lost.var
```

Now we cluster.
```{r cluster-all}
g <- buildSNNGraph(merged.pbmc, use.dimred="corrected")
colLabels(merged.pbmc) <- factor(igraph::cluster_louvain(g)$membership)
table(colLabels(merged.pbmc), merged.pbmc$batch)
```

And visualize!

```{r  viz-umap-bc}
set.seed(10101010)
merged.pbmc <- runUMAP(merged.pbmc, dimred="corrected")
gridExtra::grid.arrange(
    plotUMAP(merged.pbmc, colour_by="label", text_by="label", text_colour="red"),
    plotUMAP(merged.pbmc, colour_by="batch")
)
```

```{r  viz-tsne-bc}
set.seed(10101010)
merged.pbmc <- runTSNE(merged.pbmc, dimred="corrected")
gridExtra::grid.arrange(
    plotTSNE(merged.pbmc, colour_by="label", text_by="label", text_colour="red"),
    plotTSNE(merged.pbmc, colour_by="batch")
)
```

## 'Interpretation'

This didn't run on the merged data, so let me try it on pbmc3k.
```{r markers}
markers <- findMarkers(all.sce[[1]], pval.type="some", direction="up")
```

```{r cluster8}
marker.set <- markers[["8"]]
as.data.frame(marker.set[1:30,1:3])
```

```{r plotExpr}
plotExpression(all.sce[[1]], features=c("ENSG00000105369", "ENSG00000007312",
    "ENSG00000204287", "ENSG00000231389"), x="label", colour_by="label")
```

 Distribution of expression values for monocyte and macrophage markers across clusters in the PBMC dataset.

## SessionInfo
```{r si}
sessionInfo()
```


