---
title: "Genomewide average PWD"
author: "ks"
date: "2024-10-04"
output: html_document
---

```{r setup-nest, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Goal: Read in the beta values for a list of samples and compute pairwise distances.

```{r libraries-nest, message=FALSE}
library(GenomicFeatures)
library(AnnotationDbi)
library(rtracklayer)
library(data.table)
library(ggplot2)
```

Here is the list of sample names, their file locations and names.
```{r data-files}
samplename <- c("DA","DB","HA","HB","IA","IB","JA","JB","MA","MB","JN","IN")

readdir1 <- c("~kims/Google Drive/My Drive/Data/hpc-archive/wg_bed/bed_files")
readdir2 <- c("~kims/Google Drive/My Drive/Data/hpc-archive/wg_bed/wg_bed5963")
#readdir <- c("/Volumes/extreme ssd/Data/hpc-archive/wg_bed/wg_bed5963")

fnames <- c("DA.bed","DB.bed","HA.bed","HB.bed","IA.bed","IB.bed",
            "JA.bed","JB.bed","MA.bed","MB.bed","wg_s1c.bed","wg_s4c.bed")
n <- length(fnames)
fpath <- rep(NA,n)
for (fn in 1:10)
     fpath[fn] <- file.path(readdir1,fnames[fn])
for (fn in 11:12) 
     fpath[fn] <- file.path(readdir2,fnames[fn])
```

This is a list of regions to exclude from mapped sequencing data. This list was developed as part of the ENCODE project.
```{r get-list}
# Load the exclusion list file (for hg38, as an example)
# starting here https://github.com/Boyle-Lab/Blacklist?tab=readme-ov-file takes me
# https://www.encodeproject.org/annotations/ENCSR636HFF/
# and download https://www.encodeproject.org/files/ENCFF356LFX/
excludelist <- import("../data/ENCFF356LFX.bed")
```

The data files have a header line so I will read them in using data.table and then create a GRanges object. Then I will filter on the following conditions:  
1. Restrict to chromosomes 1-22, X, Y, on the primary assembly   
2. Remove positions that fall in exclusion regions  
3. Require >4 read depth.
```{r beta_gr}
betagrLst <- NULL

#n<-2
for (fn in c(1:n)) {
    fin <- fread(fpath[fn])

    betagr <- GRanges(seqnames = fin$V1,
                   IRanges(start = fin$V2,width=1))
                  values(betagr) <- DataFrame(
                                      beta = fin$V4,
                                      depth = fin$V5
                                    )

    desired_chromosomes <- paste0("chr",c(as.character(1:22), "X", "Y"))
    betagr <- betagr[seqnames(betagr) %in% desired_chromosomes]
    # Reset seqinfo to only include chromosomes present in the GRanges object
    betagr <- keepSeqlevels(betagr, seqlevelsInUse(betagr), pruning.mode = "coarse")
    betagr <- subsetByOverlaps(betagr, excludelist, invert = TRUE)
    betagr <- betagr[values(betagr)$depth > 4]
    betagrLst <- c(betagrLst,betagr)
}
names(betagrLst) <- samplename[1:n]
```

How many CpGs are measured in each sample after filtering?
```{r nCpGs}
lapply(betagrLst,length)
```


Now compute pairwise differences from overlaps
```{r findOverlaps}

pairtype <- c("NN","WT","BT","NT")

for (i in 1:(n-1)) {
  for (j in (i+1):n) {

      fO <- findOverlaps(betagrLst[[i]],betagrLst[[j]])
      fO

      # restrict to overlapping CpGs
      os1 <- betagrLst[[i]][queryHits(fO)]
      os2 <- betagrLst[[j]][subjectHits(fO)]
      pwd_gr <- os1
      # I could make this a data table to speed things up
      values(pwd_gr) <- data.frame( pwd = abs(values(os1)$beta - values(os2)$beta),
                                    avg =    (values(os1)$beta + values(os2)$beta)/2)
      #pwd_gr
      ptype <- ifelse(substr(samplename[i],2,2)=="N" & substr(samplename[j],2,2)=="N", 1,
                     ifelse(substr(samplename[i],1,1)==substr(samplename[j],1,1) &
                            substr(samplename[i],2,2)!="N" & substr(samplename[j],2,2)!="N",2,
                     ifelse(substr(samplename[i],1,1)!=substr(samplename[j],1,1) &
                            (substr(samplename[i],2,2)!="N" & substr(samplename[j],2,2)!="N"),3,
                            4)))
      
      
      out<- data.frame(pair = paste0(samplename[i],samplename[j]),
                       pairtype = pairtype[ptype],
                       pwd = round(mean(values(pwd_gr)$pwd),6),
                       avg = round(mean(values(pwd_gr)$avg),6),
                       nCpG = length(pwd_gr))
      #out
      fwrite(out, file = "output.txt", sep = "\t", quote = FALSE, append = TRUE)
  }
}
```

## Genomewide average PWDs

```{r summarize-pwd}
out <- fread("output.txt",sep = "\t")
head(out)
```

```{r}
p <- ggplot(out,mapping = aes(nCpG,pwd,color = pairtype)) +
           geom_point() +
           labs(color="Pair Type")
p
```

```{r}
ggplot(out, aes(x=pairtype, y = pwd, color = pairtype)) +
        geom_point() +
        stat_summary(fun=median, geom="point", shape=23,size=2, color="black")
```


```{r sI-nest}
sessionInfo()
```

