---
title: "DNA methylation profiles by gene expression variation"
author: "ks"
date: "2024-09-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE}
library(GenomicRanges)
library(data.table)
library(ggplot2)
```

# Compute Average DNAm by site across intervals


```{r load-data}
load("data/beta.by.region.rda")

Dt <- data.table(cpgpos = start(beta.by.region),
                    as.data.frame(values(beta.by.region)))
Dt$pos <- ifelse(values(beta.by.region)$tsstrand=="-",
                  values(beta.by.region)$tss -
                  start(beta.by.region),
                 start(beta.by.region) - 
                 values(beta.by.region)$tss)
head(Dt)
```

Add gene expression deviance for stratifying DNA methylation data. 
```{r merge-w-region-variables}
load("data/scdev.rda")

scdev <- data.table(scdev)
scdev <- scdev[!is.na(gene_ids)]
mDt <- merge(x = Dt,
             y = scdev,
             by.x = c("tssgene_id"),
             by.y = c("gene_ids"),
             all.y = TRUE)
mDt
```


Let's count the number of CpGs in each position, by strand.
```{r}
nobs<- table(mDt$pos,
             mDt$tsstrand)
head(nobs)
```


How many intervals are on the plus strand and how many are on the minus strand, grouped by average beta value of interval?
```{r Dt-counts}
mDt <- mDt[order(tsstrand)]
mDt[, .(unique_tss_count = uniqueN(tssgene_id)), 
              by = c("tsstrand")]
```

```{r avgbeta}
avgbeta <- mDt[, .(avg_beta = mean(beta)), 
                  by = c("tsstrand","pos")]
avgbeta <- avgbeta[order(tsstrand,pos)]
avgbeta
```

```{r}
ggplot(avgbeta) +
        geom_point(aes(x=pos,y=avg_beta,color=factor(tsstrand), alpha=0.3)) +
        geom_smooth(aes(x=pos,y=avg_beta,color=factor(tsstrand))) +
        facet_grid( ~ as.factor(tsstrand))
```

```{r sI}
sessionInfo()
```

