---
title: "Summarize CpG counts by site and regions"
author: "ks"
date: "2024-09-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE}
library(GenomicRanges)
library(data.table)
library(ggplot2)
```

# Plot number of CpGs (coverage) by position relative to TSS


```{r load-data}
load("data/beta.by.region.rda")

Dt <- as.data.table(values(beta.by.region))
Dt$pos <- ifelse(values(beta.by.region)$tsstrand=="-",
                  values(beta.by.region)$tss -
                  start(beta.by.region),
                 start(beta.by.region) - 
                 values(beta.by.region)$tss)
head(Dt)
Dt$tsstrand <- factor(Dt$tsstrand)
Dt$TSSinCGI <- factor(Dt$TSSinCGI)


Dt$group <- ifelse(Dt$pos<0, 1,
                   ifelse(Dt$pos==0, 2, 3))
head(Dt)
```

Cecilia Greenwood's work reports that beta values are correlated with read depth.
Let's check out that bias here.
```{r assoc-beta-depth}
fit <- lm(beta ~ depth , data = Dt)
summary(fit)$coef
```
How can I take correlation into account? Maybe sample just 200 genes.
```{r assoc-beta-depth}
set.seed(101)
gene.subset <- sample(unique(Dt$tssgene_id),1000)
fit <- lm(beta ~ depth + factor(tssgene_id), data = Dt[tssgene_id %in% gene.subset])
summary(fit)$coef[1:5,]
```

It looks like it could. I did get 1 result with only 200 genes (seed=101) that had p=0.3,
but when I increase the number of genes, the slight trend is statistically significant. But is this the correct model? Measures are likely correlated within strands. So what is the proper correlation structure?  Is this one equivalent to exchangeable, which is a decent approximation? Maybe if I limit to downstream of TSS. That would be interesting.


Let's count the number of CpGs in each position, by strand. I created a variable group, which takes on values 1, 2, 3 depending on whether the CpG is left or right of the TSS. 

```{r count_CpGs}
numCpGs <- Dt[, .(n_CpG = .N), by = c("tsstrand","pos")]
numCpGs <- numCpGs[order(tsstrand,pos)]
numCpGs
```

Let's plot this as the fraction of TSS intervals by strand.
```{r Dt-counts}
nI <- Dt[, .(unique_tss_count = uniqueN(tssgene_id)), by = tsstrand]
nI
```

```{r}
Df <- merge(numCpGs,nI)
Df$freq_CpG <- Df$n_CpG/Df$unique_tss_count
Df
```


```{r}
ggplot(Df) +
        geom_point(aes(x=pos,y=freq_CpG,color=tsstrand), alpha=0.3) +
        geom_smooth(aes(x=pos,y=freq_CpG,color=tsstrand)) +
        facet_wrap(~tsstrand)
```

Although computed from the data (5+ reads), this suggests there are fewer CpGs per position upstream of the TSS, which is consistent with lower CpG density regions and consequently higher DNA methylation. CpG density can confound any direct comparison of average DNA methylation between regions with different CpG density.


Let's compute 3 summary statistics for each interval: 
1) number upstream CpGs 
2) TSS has CpG (1/0)  
3) number of downstream CpGs
```{r count_CpGs_byregion}
Dt <- Dt[order(tsstrand,tssgene_id,group)]
numCpGs <- Dt[, .(n_CpG = .N, 
                  avg_beta = mean(beta), 
                  sd_depth = sd(depth)),
                   by = c("tsstrand","tssgene_id","group")]
numCpGs
```

average Beta is negatively associated with number of CpGs in 500 bp region
```{r}
cor.test(numCpGs$n_CpG,numCpGs$avg_beta)
```

```{r}
#GGally::ggpairs(numCpGs[, tssgene_id:=NULL])
cor.test(numCpGs$sd_depth,numCpGs$avg_beta)
```

```{r}
cor.test(numCpGs$sd_depth,numCpGs$n_CpG)
```

All these associations support regions with fewer CpGs, have higher average DNA methylation and more variation in read depth. These can confound the association of average DNA methylation by CpG position relative to TSS.

```{r}
table(numCpGs$group,numCpGs$tsstrand)
```

The fact that the sample size differs upstream and downstream shows that not all strands contribute data to both regions.
Any comparisons of upstream to downstream DNA methylation, should at least match on TSS region. 

```{r}
dt_filtered <- numCpGs[, .SD[.N == 3], by = tssgene_id]

# View the filtered data.table
dt_filtered
```

```{r}
table(dt_filtered$group,dt_filtered$tsstrand)
```


```{r}
ggplot(dt_filtered, aes(x=as.factor(group),y=n_CpG))  +
        geom_boxplot() +
        facet_wrap(~tsstrand) +
        ggtitle("Regions with CpG at TSS: (n=334 for + strand, n=1498 for - strand.  
                Group = upstream, TSS, downstream")
```

Try making the figure for regions without TSS covered, but having data in both upstream and downstream regions.
```{r}

dt_filtered <-  numCpGs[ group !=2 ] 
dt_filtered <-  dt_filtered[, .SD[.N == 2], by = tssgene_id]

# View the filtered data.table
dt_filtered
```

```{r}
table(dt_filtered$group,dt_filtered$tsstrand)
```

```{r}
ggplot(dt_filtered, aes(x=as.factor(group),y=n_CpG))  +
        geom_boxplot() +
        facet_wrap(~tsstrand) +
        ggtitle("Regions with CpGs both up and downstream: (n=15198 for + strand, n=14545 for - strand.  
                Group = upstream, TSS, downstream")
```

```{r}
t.test(n_CpG ~ as.factor(group), subset = tsstrand =="+", data=dt_filtered)
```

```{r}
t.test(n_CpG ~ as.factor(group), subset = tsstrand =="-", data=dt_filtered)
```

The variation in depth makes me wonder about computing unweighted vs weighted average of betas in regions with fewer CpGs.  

```{r sI}
sessionInfo()
```

