---
title: "PWD profiles by gene expression variation"
author: "ks"
date: "2024-09-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE}
library(GenomicRanges)
library(data.table)
library(ggplot2)
```

# Compute Average DNAm by site across intervals

```{r load-data}
load("data/pwd.by.region.rda")
seqlevels(pwd.by.region)
```

Check that I dropped X and Y chromosomes and compute position of CpG relative to TSS in a strand aware fashion.
```{r data-table}
Dt <- data.table(cpgpos = start(pwd.by.region),
                           data.frame(values(pwd.by.region)))
Dt$pos <- ifelse(values(pwd.by.region)$tsstrand=="-",
                  values(pwd.by.region)$tss -
                  start(pwd.by.region),
                 start(pwd.by.region) - 
                 values(pwd.by.region)$tss)
head(Dt)
Dt$tsstrand <- factor(Dt$tsstrand)
Dt$TSSinCGI <- factor(Dt$TSSinCGI)
```

Add gene expression deviance for stratifying DNA methylation data. 
```{r merge-w-genedev}
load("data/scdev.rda")

scdev <- data.table(scdev)
scdev <- scdev[!is.na(gene_ids)]
mDt <- merge(x = Dt,
             y = scdev,
             by.x = c("tssgene_id"),
             by.y = c("gene_ids"))
mDt
```

Add interval average beta for stratifying DNA methylation data. 
```{r merge-w-region-variables}
load("data/filt_sum.tssregion.rda")

mDt <- merge(x = mDt,
             y = filt_sum.tssregion)
mDt
```

Let's count the number of CpGs in each position, by strand.
```{r}
nobs<- table(mDt$pos,
             mDt$tsstrand)
head(nobs)
```

## Stratify by interval averaged DNAm level

How many TSSs are in or not in CpG islands, grouped by average beta value of interval?
```{r Dt-counts}
mDt <- mDt[order(TSSinCGI,cat_region_avg_beta)]
mDt[, .(unique_tss_count = uniqueN(tssgene_id)), 
              by = c("TSSinCGI","cat_region_avg_beta")]
```

```{r avgbeta}
mDt <- mDt[order(TSSinCGI,cat_region_avg_beta,pos)]
avgpwd <- mDt[, .(avg_pwd = mean(pwd)), 
                  by = c("TSSinCGI","cat_region_avg_beta","pos")]
avgpwd <- avgpwd[order(TSSinCGI,cat_region_avg_beta,pos)]
```

```{r}
ggplot(avgpwd) +
        geom_point(aes(x=pos,y=avg_pwd,color=cat_region_avg_beta), alpha=0.3) +
        geom_smooth(aes(x=pos,y=avg_pwd,color=cat_region_avg_beta)) +
        facet_grid( ~ TSSinCGI)
```

## Stratify by gene expression deviance

Now summarize by gene expression deviance instead of average methylation level.

```{r Dt-counts-catgexdev}
ldev <- quantile(mDt$colon_ldev, probs = seq(0, 1, by = 0.25))
mDt$cat_colon_ldev <- factor(cut(mDt$colon_ldev, breaks = ldev, include.lowest = TRUE, labels = FALSE))

mDt <- mDt[order(TSSinCGI,cat_colon_ldev)]
mDt[, .(unique_tss_count = uniqueN(tssgene_id)), 
              by = c("TSSinCGI","cat_colon_ldev")]
```

```{r avgpwd-scdev}
mDt <- mDt[order(TSSinCGI,cat_colon_ldev,pos)]
avgpwd <- mDt[, .(avg_pwd = mean(pwd)), 
                  by = c("TSSinCGI","cat_colon_ldev","pos")]
avgpwd <- avgpwd[order(TSSinCGI,cat_colon_ldev,pos)]
```

```{r}
ggplot(avgpwd) +
        geom_point(aes(x=pos,y=avg_pwd,color=cat_colon_ldev), alpha=0.3) +
        geom_smooth(aes(x=pos,y=avg_pwd,color=cat_colon_ldev)) +
        facet_grid( ~ TSSinCGI)
```


## Stratify by number of CpGs in interval (instead of avg beta)




```{r sI}
sessionInfo()
```

