---
title: "PWD profiles by CTCF"
author: "ks"
date: "2024-09-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE}
library(GenomicRanges)
library(data.table)
library(ggplot2)
library(knitr)
library(kableExtra)
```


## Get CTCF intervals

```{r read-ctcf-motifs}
load("data/ctcf_motifs.rda")
values(ctcf_motifs)$ctcf_idx <- 1:length(ctcf_motifs)
```

We've downloaded a total of `r length(ctcf_motifs)` CTCF motif intervals from https://pubmed.ncbi.nlm.nih.gov/39236169/ (Science 2024 - scATACseq in TCGA)

```{r kable, echo=FALSE}
tb <- table(values(ctcf_motifs)$cancer_type)
kable(tb, col.names = c("Cancer Type", "Count")) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```


```{r summarize-widths, echo=FALSE}
widthdistn <- cbind(summary(width(ctcf_motifs[values(ctcf_motifs)$cancer_type=="BRCA"])),
                    summary(width(ctcf_motifs[values(ctcf_motifs)$cancer_type=="COAD"])),
      summary(width(ctcf_motifs[values(ctcf_motifs)$cancer_type=="COAD+BRCA"])))
kable(widthdistn, col.names = c("BRCA","COAD","COAD+BRCA"),
           caption = c("Distribution of Interval Widths")) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```


Summarize average DNAm and PWD in these intervals.   
Average DNAm over entire interval.  Show Violin plot by # CpGs  

## Get Normal PWDs

### Sample: IA

```{r load-JN}
filtered_data <- readRDS("~kims/github/sandbox/bsseq/data/filtered_data.rds")
jn <- filtered_data[values(filtered_data)$sample=="IA"]
```

These motif intervals are narrow, so I'm going to pool reads in the interval to estimate the pooled beta value. I will process each sample separately before merging samples to compute PWD.
 
```{r jn-motif-interval}
fO <- findOverlaps(ctcf_motifs, jn)
#fO
```

`r length(fO)` CpG measures map to `r length(unique(queryHits(fO)))` CTCF intervals.

```{r subset-and-merge-values}
jn <- jn[subjectHits(fO)]
values(jn)$ctcf_idx <- values(ctcf_motifs)$ctcf_idx[queryHits(fO)]
values(jn)$pos <- start(jn)
jn
```

Summarize number of CpGs and the beta value by CTCF motif interval.
```{r summarize-jn-by-ctcfmotif}
Dt <- as.data.table(values(jn))
Dt$filtbeta <- ifelse(Dt$depth>4,Dt$beta,NA)
Dt$lowdepth <- ifelse(Dt$depth<5,1,0)

Dt <- Dt[order(ctcf_idx)]
sDt <- Dt[,.( nCpG = .N,
           avg_filtbeta = mean(filtbeta,na.rm=T),
           uncovered = sum(lowdepth),
           summ     = sum(m),
           sumdepth = sum(depth)), by = ctcf_idx]
sDt$poolbeta <- sDt$summ/sDt$sumdepth

#Add annotation of cancer_type for each interval: COAD,COAD+BRCA,BRCA
idx <- which(is.element(values(ctcf_motifs)$ctcf_idx,sDt$ctcf_idx))
sDt$cancer_type <- values(ctcf_motifs)$cancer_type[idx]
```

#### DNAm by CTCF region/nCpGs

N = `r nrow(sDt)` intervals (`r round(nrow(sDt)/length(ctcf_motifs)*100,1)`$\%$) have DNA methylation measurements. Of these, `r round(mean(sDt$sumdepth>4)*100,1)`$\%$ have 5 or more reads.  Let's count the number of intervals containing 1 - 6 CpGs.

```{r kable-cpgcount, echo=FALSE}
# didn't subtract number uncovered
tb <- table(sDt$nCpG[sDt$sumdepth>4],sDt$cancer_type[sDt$sumdepth>4])
kable(tb, col.names = c("Number of CpGs", "BRCA","COAD","COAD+BRCA")) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

Does average DNA methylation in the CTCF motif interval vary depending on which cancer type had the predicted accessible TFBS?

```{r smoothed-curve}
sDt$numCpG <- factor(ifelse(sDt$nCpG > 5, 6, sDt$nCpG),
                         labels = c("1","2","3","4","5","6+"))
ggplot(sDt[sumdepth>4], aes(x=cancer_type, y = poolbeta, fill = cancer_type)) +
        geom_violin() +
  # violin plot with mean points
        stat_summary(fun=mean, geom="point", shape=23, size=2) +
# violin plot with median points
        stat_summary(fun=median, geom="point", size=2, color="red") +
        facet_wrap( ~ factor(numCpG))
```

Colored diamonds are means. Red points are medians.  

```{r lm-jn}
sDt$shiftnumCpG <- as.numeric(sDt$numCpG) - 1
summary(lm(poolbeta ~ cancer_type + shiftnumCpG, data=sDt, subset=sumdepth > 4))
```

Let's get the second normal tissue sample and summarize PWD for these locations.
I might include different CpGs by doing this.

### Sample: IB

```{r load-IN}
sampleIN <- filtered_data[values(filtered_data)$sample=="IB"]
fO <- findOverlaps(ctcf_motifs, sampleIN)
sampleIN <- sampleIN[subjectHits(fO)]
values(sampleIN)$ctcf_idx <- values(ctcf_motifs)$ctcf_idx[queryHits(fO)]
values(sampleIN)$pos <- start(sampleIN)
sampleIN <- sampleIN[order(sampleIN)]
```

`r length(fO)` CpG measures map to `r length(unique(queryHits(fO)))` CTCF intervals.

Summarize number of CpGs and the beta value by CTCF motif interval.
```{r summarize-in-by-ctcfmotif}
Dti <- as.data.table(values(sampleIN))
Dti$filtbeta <- ifelse(Dti$depth>4,Dti$beta,NA)
Dti$lowdepth <- ifelse(Dti$depth<5,1,0)

Dti <- Dti[order(ctcf_idx)]
sDti <- Dti[,.( nCpG = .N,
           avg_filtbeta = mean(filtbeta,na.rm=T),
           uncovered = sum(lowdepth),
           summ     = sum(m),
           sumdepth = sum(depth)), by = ctcf_idx]
sDti$poolbeta <- sDti$summ/sDti$sumdepth

#Add annotation of cancer_type for each interval: COAD,COAD+BRCA,BRCA
idx <- which(is.element(values(ctcf_motifs)$ctcf_idx,sDti$ctcf_idx))
sDti$cancer_type <- values(ctcf_motifs)$cancer_type[idx]
```


#### DNAm by CTCF region/nCpGs

N = `r nrow(sDti)` intervals (`r round(nrow(sDti)/length(ctcf_motifs)*100,1)`$\%$) have DNA methylation measurements. Of these, `r round(mean(sDti$sumdepth>4)*100,1)`$\%$ have 5 or more reads.  Let's count the number of intervals with 5+ reads by the number of CpGs in the interval.

```{r kable-cpgcounti, echo=FALSE}
tb <- table(sDti$nCpG[sDti$sumdepth>4],sDti$cancer_type[sDti$sumdepth>4])
kable(tb, col.names = c("Number of CpGs", "BRCA","COAD","COAD+BRCA")) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```

Does average DNA methylation in the CTCF motif interval vary depending on which cancer type had the predicted accessible TFBS?

```{r pooledbeta-distn-sample2}
sDti$numCpG <- factor(ifelse(sDti$nCpG > 5, 6, sDti$nCpG),
                         labels = c("1","2","3","4","5","6+"))
ggplot(sDti[sumdepth>4], aes(x=cancer_type, y = poolbeta, fill = cancer_type)) +
        geom_violin() +
  # violin plot with mean points
        stat_summary(fun=mean, geom="point", shape=23, size=2) +
# violin plot with median points
        stat_summary(fun=median, geom="point", size=2, color="red") +
        facet_wrap( ~ factor(numCpG))
```

Colored diamonds are means. Red points are medians.  

```{r lm-in}
sDti$shiftnumCpG <- as.numeric(sDti$numCpG) - 1
summary(lm(poolbeta ~ cancer_type + shiftnumCpG, data=sDti, subset=sumdepth > 4))
```

## Merge and plot PWDs

Merge the interval summary stats and compute PWD.

Change column names for the second sample so I can merge and 
compute pairwise differences.
```{r chg-colnames}
cn <- colnames(sDti)
cn[c(2:7,9,10)] <- paste0("s2.",cn[c(2:7,9,10)])
cn
colnames(sDti) <- cn
```

Compute pairwise differences
```{r sample-pair, echo=FALSE}
pairsDt <- merge(sDt,
                 sDti)
dim(pairsDt)
pairsDt$pwd <- abs(pairsDt$poolbeta - pairsDt$s2.poolbeta)
```

```{r PWD}
ggplot(pairsDt[sumdepth>4 & s2.sumdepth>4], aes(x=cancer_type, y = pwd, fill = cancer_type)) +
        geom_violin() +
  # violin plot with mean points
        stat_summary(fun=mean, geom="point", shape=23, size=2) +
# violin plot with median points
        stat_summary(fun=median, geom="point", size=2, color="red") +
        ylim(0,.4) + 
        facet_wrap( ~ factor(numCpG))
```

```{r lm-pwd}
ft <- pairsDt$sumdepth > 4 & pairsDt$s2.sumdepth > 4
summary(lm(pwd ~ cancer_type + shiftnumCpG, data=pairsDt, subset=ft))
```

```{r sI}
sessionInfo()
```

