---
title: "DNA methylation profiles"
author: "ks"
date: "2024-11-06"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries, message=FALSE}
library(GenomicRanges)
library(data.table)
library(ggplot2)
library(rtracklayer)
samplename <- c("DA","DB")
```

# Summarize data for matched samples: `r samplename`

First, read in the data for samples.

```{r data-dir}
# this is for the normal samples
#readdir <- c("~kims/Google Drive/My Drive/Data/hpc-archive/wg_bed/wg_bed5963")
#fin1 <- fread(file.path(readdir,"wg_s1c.bed"))
#fin1$sample <- samplename[1]

#fin2 <- fread(file.path(readdir,"wg_s4c.bed"))
#fin2$sample <- samplename[2]
#readdir <- c("/Volumes/extreme ssd/Data/hpc-archive/wg_bed/wg_bed5963")

# This is for paired tumor samples
readdir <- c("~kims/Google Drive/My Drive/Data/hpc-archive/wg_bed/bed_files")
fin1 <- fread(file.path(readdir,paste0(samplename[1],".bed")))
fin1$sample <- samplename[1]

fin2 <- fread(file.path(readdir,paste0(samplename[2],".bed")))
fin2$sample <- samplename[2]
```


```{r read-beta-values, echo=FALSE}
combined_fin <- rbindlist(list(fin1, fin2))
setnames(combined_fin, c("chr","start","end","beta","depth","m","u","sample"))
rm(fin1,fin2)

desired_chromosomes <- paste0("chr",c(as.character(1:22),"X","Y"))
combined_fin <- combined_fin[chr %in% desired_chromosomes]

combined_fin[, .(n_CpG = .N, 
              avg_beta = mean(beta), 
           median_beta = median(beta),
             avg_depth = mean(depth),
             max_depth = max(depth),
           frac_depth5 = mean(depth>4)), by = sample]
```

Notice the high maximum read depth.  This reminds me there are regions of the genome that should be excluded due to artifacts.  Let's go find the list from ENCODE.

## Remove Exclusion list regions
Here's how ChatGPT tells me to filter these regions:

```{r get-list}
# Load the exclusion list file (for hg38, as an example)
# starting here https://github.com/Boyle-Lab/Blacklist?tab=readme-ov-file takes me
# https://www.encodeproject.org/annotations/ENCSR636HFF/
# and download https://www.encodeproject.org/files/ENCFF356LFX/
excludelist <- import("data/ENCFF356LFX.bed")
```

Remove regions.

```{r make-GRobject}
betavals <- GRanges(seqnames = combined_fin$chr,
                   IRanges(start = combined_fin$start,width=1))
values(betavals) <- DataFrame(beta = combined_fin$beta,
                             depth = combined_fin$depth,
                                 m = combined_fin$m,
                                 u = combined_fin$u,
                            sample = combined_fin$sample)

filtered_data <- subsetByOverlaps(betavals, excludelist, invert = TRUE)
excluded_data <- subsetByOverlaps(betavals, excludelist)
```


Now summarize the remaining data.
```{r make-Dt}
Dt <- as.data.table(values(filtered_data))
Dt[, .(n_CpG = .N, 
           avg_beta = mean(beta), 
        median_beta = median(beta),
          avg_depth = mean(depth),
       median_depth = as.numeric(median(depth)),
          max_depth = max(depth),
       frac_depth5 = mean(depth>4)), by = sample]
```
How does this compare to the excluded data?
```{r sum-excluded-data}
Dt2 <- as.data.table(values(excluded_data))
Dt2[, .(n_CpG = .N, 
           avg_beta = mean(beta), 
        median_beta = median(beta),
          avg_depth = mean(depth),
       median_depth = median(depth),
          max_depth = max(depth)), by = sample]
```

`r round(length(excluded_data)/length(betavals)*100,1)`$\%$ of measures were excluded.
That was a principled way to remove some high read depths, but it didn't get them all.


```{r save}
fn <- paste0("data/filtered_samples",samplename[1],samplename[2],".rds")
saveRDS(filtered_data,file=fn)
```

## Characterize the data

Do this for sample `r samplename[1]`. The tumor data appear even more variable with respect to sequencing depth across the genome.  I will remove the high read depths before studying the relationship between beta and depth. Run regressions for sample `r samplename[1]`.

### Sample `r samplename[1]`

```{r readdepth, echo=FALSE}
Dt$logdepth <- log(Dt$depth+1)
ggplot(Dt[which(sample %in% samplename[1] & seqnames(filtered_data)=="chr22")], 
          aes(x=logdepth, y = beta) ) +
          geom_point() +
          geom_smooth(span=0.5) +
          geom_smooth(method="lm",color="red") +
          labs(title=paste0(samplename[1],": chr22"))
```

```{r lmfit-beta-filtered}
summary(lm(beta ~ depth,data=Dt, 
                        subset = sample == samplename[1]))$coef
```

```{r lmfit-beta-filtered-depth}
summary(lm(beta ~ depth,data=Dt, 
           subset = sample == samplename[1] & depth < 400 ))$coef
```
```{r muplot, echo=FALSE}
ggplot(Dt[which(sample %in% samplename[1] & seqnames(filtered_data)=="chr22")], 
          aes(x=u, y = m) ) +
          geom_point() +
          geom_smooth(span=0.5) +
          geom_smooth(method="lm",color="red") +
          labs(title=paste0(samplename[1],": chr22"))
```


```{r lmfit-m-filtered}
summary(lm(m ~ u, data=Dt, subset = sample==samplename[1]))$coef
```

```{r lmfit-m-depth-cutoff}
summary(lm(m ~ u, data=Dt, subset = sample ==samplename[1] & depth < 400))$coef
```

### Sample `r samplename[2]`

```{r readdepth-s2, echo=FALSE}
Dt$logdepth <- log(Dt$depth+1)
ggplot(Dt[which(sample %in% samplename[2] & seqnames(filtered_data)=="chr22")], 
          aes(x=logdepth, y = beta) ) +
          geom_point() +
          geom_smooth(span=0.5) +
          geom_smooth(method="lm",color="red") +
          labs(title=paste0(samplename[2],": chr22"))
```

```{r lmfit-beta-filtered-s2}
summary(lm(beta ~ depth,data=Dt, 
                        subset = sample == samplename[2]))$coef
```

```{r lmfit-beta-filtered-s2-depth}
summary(lm(beta ~ depth,data=Dt, 
                        subset = sample == samplename[2] & depth < 400))$coef
```

```{r muplot-s2, echo=FALSE}
ggplot(Dt[which(sample %in% samplename[2] & seqnames(filtered_data)=="chr22")], 
          aes(x=u, y = m) ) +
          geom_point() +
          geom_smooth(span=0.5) +
          geom_smooth(method="lm",color="red") +
          labs(title=paste0(samplename[2],": chr22"))
```


```{r lmfit-m-filtered-s2}
summary(lm(m ~ u, data=Dt, subset = sample==samplename[2]))$coef
```

```{r lmfit-m-depth-cutoff-s2}
summary(lm(m ~ u, data=Dt, subset = sample ==samplename[2] & depth < 400))$coef
```
Even stronger effect when filtering depth $<$ 400.  But this relies on an arbitrary cutoff so we do not exclude them from the saved data set.


```{r sI}
sessionInfo()
```

