---
title: "DNAm summaries"
author: "ks"
date: "2024-09-27"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries}
library(GenomicFeatures)
library(AnnotationDbi)
library(rtracklayer)
#BiocManager::install("TxDb.Hsapiens.UCSC.hg38.knownGene")
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(data.table)
```

TASK: Overlap Sample DNAm values with the TSS regions and compute average DNAm at each site by stran, and plot it.   

1. Load the saved TSSs and extend interval to 500 bases up- and downstream.
```{r tss-data}
load("min_tss_gr")
min_tss_gr <- sort(min_tss_gr)
# Resize to extend 500 bases upstream and downstream (i.e., total width of 1001 bp)
tss_resized <- resize(min_tss_gr, width = 1001, fix = "center")
```

2. Now grab a DNA methylation dataset. The file below
is setup to read sample IN (aka N2) and save all CpGs that have coverge of 5 or more reads.
```{r dnam-data}
rmarkdown::render("read-dnam-bedfiles.Rmd")
```

We're starting with `r length(betan2)` measured CpGs.
  
3. Try overlapping my 2 GRange objects. 

One thing I discovered is that I have overlapping TSS intervals.  Should I reduce them to non-overlapping intervals or keep their full overlapping lengths and allow some CpGs to map to multiple intervals? I will try the latter.


```{r overlap}
tss_resized <- sort(tss_resized)
fO <- findOverlaps(betan2,tss_resized)
fO
```
number of CpGs in intervals: `r length(fO)`
number of unique CpGs in intervals: `r length(unique(queryHits(fO)))`

How are these distributed across different TSS intervals?

```{r describe-fO}
length(fO)

table(table(queryHits(fO)))
# Split the DNAm data by TSSs
#betan2_by_tss <- split(tss_gr, tss_gr$gene_id)
```

953369 are in 1 TSS interval  
163955 are in 2 TSS intervals  
 12807 are in 3 TSS intervals. 
 ...  
 
Can I save the queryHits(fO) rows of betan2? First, I want to give my CpGs an cpgid. I will get do this in 
sequence order.

```{r filt-cpgs}
betan2 <- betan2[queryHits(fO)]
filt.betan2
```


```{r filt-cpgs}
filt.betan2 <- betan2[queryHits(fO)]
filt.betan2
```


```{r sI}
sessionInfo()
```

